// @ovine/cli/lib/webpack/babel.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
exports.__esModule = true;
exports.getDllBabelConfig = exports.getBabelConfig = void 0;
/**
 * babel-loader configuration
 */
var fs_extra_1 = __importDefault(require("fs-extra"));
var lodash_1 = require("lodash");
var path_1 = __importDefault(require("path"));
var constants_1 = require("../constants");
var _a = process.env.NODE_ENV, NODE_ENV = _a === void 0 ? 'development' : _a;
var styledDefConf = {
    development: {
        displayName: true
    },
    production: {
        minify: true,
        pure: true,
        displayName: false
    }
};
// function importPlugin(moduleName, dirName = '') {
//   return [
//     'babel-plugin-import',
//     {
//       libraryName: moduleName,
//       libraryDirectory: '',
//       camel2DashComponentName: false,
//     },
//     dirName,
//   ]
// }
function extendsConfig(siteDir) {
    var configFile = path_1["default"].resolve(siteDir, constants_1.babelConfigFileName);
    if (!fs_extra_1["default"].existsSync(configFile)) {
        return {};
    }
    return {
        "extends": configFile
    };
}
function getBabelConfig(option) {
    var siteDir = option.siteDir, _a = option.styledConfig, styledConfig = _a === void 0 ? {} : _a;
    var styledConf = lodash_1.defaultsDeep(styledConfig, styledDefConf);
    var styledFinalConf = __assign(__assign({}, lodash_1.omit(styledConf, ['development', 'production'])), lodash_1.get(styledConf, NODE_ENV));
    return __assign(__assign({}, extendsConfig(siteDir)), { compact: false, presets: [
            [
                '@babel/preset-env',
                {
                    targets: {
                        esmodules: true
                    }
                },
            ],
            '@babel/preset-react',
        ], plugins: [
            'react-hot-loader/babel',
            'babel-plugin-macros',
            ['babel-plugin-styled-components', styledFinalConf],
            '@babel/plugin-syntax-dynamic-import',
        ] });
}
exports.getBabelConfig = getBabelConfig;
function getDllBabelConfig(siteDir) {
    return __assign(__assign({}, extendsConfig(siteDir)), { compact: false, plugins: ['@babel/plugin-syntax-dynamic-import'] });
}
exports.getDllBabelConfig = getDllBabelConfig;
