// @ovine/cli/lib/config.js
"use strict";
/**
 * load & check site config
 */
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __spreadArrays = (this && this.__spreadArrays) || function () {
    for (var s = 0, i = 0, il = arguments.length; i < il; i++) s += arguments[i].length;
    for (var r = Array(s), k = 0, i = 0; i < il; i++)
        for (var a = arguments[i], j = 0, jl = a.length; j < jl; j++, k++)
            r[k] = a[j];
    return r;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
exports.__esModule = true;
exports.loadContext = exports.loadConfig = void 0;
var fs_extra_1 = __importDefault(require("fs-extra"));
var import_fresh_1 = __importDefault(require("import-fresh"));
var lodash_1 = __importDefault(require("lodash"));
var path_1 = __importDefault(require("path"));
var constants_1 = require("./constants");
var requiredFields = ['favicon', 'title'];
var optionalFields = [
    'publicPath',
    'dll',
    'envModes',
    'devServer',
    'ui',
    'staticFileExts',
    'template',
    'styledConfig',
    'cacheGroups',
    'splitRoutes',
];
var defaultConfig = {
    publicPath: '/',
    template: {},
    devServer: {},
    dll: {},
    ui: {
        defaultTheme: 'default',
        withoutPace: false
    }
};
function formatFields(fields) {
    return fields.map(function (field) { return "'" + field + "'"; }).join(', ');
}
function checkPath(pathStr, value, egText) {
    if (value === void 0) { value = ''; }
    if (value && !(typeof value === 'string' && value.substr(-1) === '/')) {
        throw new Error(pathStr + ": \"" + value + "\" is not allowed. The \"" + pathStr + "\" must be string end with \"/\". eg: \"" + (egText ||
            'https://abc.com/') + "\"");
    }
}
function loadConfig(siteDir, options) {
    var configPath = path_1["default"].resolve(siteDir, constants_1.configFileName);
    if (!fs_extra_1["default"].existsSync(configPath)) {
        throw new Error(configPath + " not found! Please check the command line args.");
    }
    var configFile = import_fresh_1["default"](configPath);
    var loadedConfig = lodash_1["default"].isPlainObject(configFile)
        ? configFile
        : lodash_1["default"].isFunction(configFile)
            ? configFile(options)
            : {};
    var missingFields = requiredFields.filter(function (field) { return !lodash_1["default"].has(loadedConfig, field); });
    if (missingFields.length > 0) {
        throw new Error("The required field(s) " + formatFields(missingFields) + " are missing from " + constants_1.configFileName);
    }
    // Merge default config with loaded config.
    var config = __assign(__assign({}, defaultConfig), loadedConfig);
    // Don't allow unrecognized fields.
    var allowedFields = __spreadArrays(requiredFields, optionalFields);
    var unrecognizedFields = Object.keys(config).filter(function (field) { return !allowedFields.includes(field); });
    var _a = config.publicPath, publicPath = _a === void 0 ? '/' : _a, dll = config.dll, envModes = config.envModes;
    checkPath('publicPath', publicPath);
    checkPath('dll.publicPath', dll.publicPath);
    checkPath('dll.hostDir', dll.hostDir);
    // TODO: use json schema for Configuration verification!
    if (unrecognizedFields.length) {
        throw new Error("The field(s) " + formatFields(unrecognizedFields) + " are not recognized in " + constants_1.configFileName);
    }
    if (envModes) {
        if (!lodash_1["default"].isArray(envModes) || envModes.some(function (i) { return typeof i !== 'string'; })) {
            throw new Error("envModes: \"" + envModes + "\" is not allowed. The \"envModes\" must be array of string. eg: [\"local\", \"test\", \"prod\"]");
        }
    }
    return config;
}
exports.loadConfig = loadConfig;
function loadContext(siteDir, options) {
    var genDir = path_1["default"].resolve(siteDir, constants_1.generatedDirName);
    var siteConfig = loadConfig(siteDir, options);
    var outDir = path_1["default"].resolve(siteDir, constants_1.outDirName);
    var srcDir = path_1["default"].resolve(siteDir, constants_1.srcDirName);
    var publicPath = siteConfig.publicPath;
    return {
        siteDir: siteDir,
        genDir: genDir,
        siteConfig: siteConfig,
        outDir: outDir,
        srcDir: srcDir,
        publicPath: publicPath
    };
}
exports.loadContext = loadContext;
