// @ovine/cli/lib/constants.js
"use strict";
// lib
exports.__esModule = true;
exports.babelConfigFileName = exports.tsConfFileName = exports.tsLintConfFileName = exports.esLintFileName = exports.webpackDllConfFileName = exports.webpackConfFileName = exports.dllJsdelivrHostDir = exports.dllAssetsFile = exports.dllManifestFile = exports.dllDirPath = exports.dllVendorDirPath = exports.dllDirName = exports.dllFileKeys = exports.dllChunkFilePrefix = exports.dllVendorFileName = exports.cssAssetsFile = exports.cssAssetsName = exports.scssDirName = exports.stylesDirName = exports.staticLibDirPath = exports.generatedDirName = exports.configFileName = exports.staticDirName = exports.outDirName = exports.srcDirName = exports.defaultPort = exports.domain = exports.winConst = exports.dllVer = exports.amisEditorVer = exports.libVer = exports.libRootPath = exports.libName = void 0;
exports.libName = 'ovine';
exports.libRootPath = "/" + exports.libName;
exports.libVer = require('../package.json').version;
exports.amisEditorVer = '2.0.14';
// publish_mark 如果当前升级不需要更新 dll,需要指定特定DLL版本,默认和版本号一致
exports.dllVer = exports.libVer;
exports.winConst = {
    dllPath: 'OVINE_DLL_PATH',
    dllVersion: 'OVINE_DLL_VERSION',
    dllRequireVer: 'OVINE_DLL_REQUIRE_VER'
};
exports.domain = {
    jsdelivr: 'https://cdn.jsdelivr.net/',
    libDoc: 'https://ovine.igroupes.com/'
};
// basic
exports.defaultPort = 7050;
exports.srcDirName = 'src';
exports.outDirName = 'dist';
exports.staticDirName = 'static';
exports.configFileName = exports.libName + ".config.js";
exports.generatedDirName = "." + exports.libName;
exports.staticLibDirPath = exports.staticDirName + "/" + exports.libName;
// scss
exports.stylesDirName = 'styles';
exports.scssDirName = 'scss';
exports.cssAssetsName = 'css_assets';
exports.cssAssetsFile = exports.generatedDirName + "/" + exports.cssAssetsName + ".json";
// dll config
exports.dllVendorFileName = 'dll_entry';
exports.dllChunkFilePrefix = 'dll_chunk_';
exports.dllFileKeys = ['boot', 'amis', exports.dllVendorFileName];
exports.dllDirName = "dll/" + exports.dllVer;
exports.dllVendorDirPath = exports.staticLibDirPath + "/" + exports.dllDirName;
exports.dllDirPath = exports.generatedDirName + "/" + exports.staticDirName + "/" + exports.dllDirName;
exports.dllManifestFile = exports.generatedDirName + "/[name]_manifest.json";
exports.dllAssetsFile = exports.generatedDirName + "/[name]_assets.json";
exports.dllJsdelivrHostDir = exports.domain.jsdelivr + "npm/@" + exports.libName + "/init@" + exports.dllVer + "/env/" + exports.dllDirPath + "/";
// config files
exports.webpackConfFileName = 'webpack.config.js';
exports.webpackDllConfFileName = 'webpack.dll.js';
exports.esLintFileName = '.eslintrc.js';
exports.tsLintConfFileName = 'tslint.json';
exports.tsConfFileName = 'tsconfig.json';
exports.babelConfigFileName = 'babel.config.js';
